### DISCUSSION TITLE
Automatically add f in beginning of string when adding curly braces within a string statement (i.e. creating f-strings)

### DISCUSSION THREAD

**Discussion by paxcodes** [microsoft/pylance-release#2615] [Enhancement]
Similar to PyCharm's [feature](https://twitter.com/i/status/1291780907343503361).

This can be an option that the user can disable.

**Comment by judej** [comment]
If this is going to be auto populate, then it should be configurable.

**Comment by ghost** [comment]
This has been finally implemented in the 2023.4.41 release and I wanted to thank the team for these small but yet time saving features! Keep it up! ;)

PS: It requires the `python.analysis.autoFormatStrings` to be enabled, hopefully they move it to the default settings in future releases.

**Reply by zach-m** [reply]
just one comment - when a sub-text inside a quoted string is first selected and then curly-braced (which adds both braces) - the "f" isn't prepended.. this is a common scenario, i believe

**Comment by pewscorner** [comment]
If it becomes the default, it is going to be an unpleasant surprise to those who create template strings for later (i.e. not immediate) use with str.format(), and to those who use '{}' in regular expressions, especially if the f-ification also happens to raw strings (r'...').

**Reply by rchiodo** [reply]
Raw strings seems like a bug. Seems like we could also skip the case were `.format` was already on the end.

**Reply by pewscorner** [reply]
The PyCharm feature seems to handle the case where .format is already at the end. But I was referring to the case where it is applied later.

**Reply by rchiodo** [reply]
I don't think that would be predictable, so if that bothered somebody they'd have to turn off the setting.

**Reply by pewscorner** [reply]
That was the point I was trying make. 😊 And because this is going to annoy people, I would argue that it should not be enabled by default. Auto-modifications should be opt-in, in my opinion.

**Reply by JeroSQ** [reply]
Hey. I use LaTeX strings in python for my job and this made me lose a lot of time. Can someone please tell me how to disable `python.analysis.autoFormatStrings` ? Thanks.

Edit: Nevermind I just figured it out, but I still don't think this should be default.

**Reply by rchiodo** [reply]
@jerosq are you saying it was on by default? Oh, maybe the experiment service turned it on. VS code runs a service that allows us to turn settings on and off for certain groups of people so we can see the impact before we default it one way or another.

**Reply by JeroSQ** [reply]
Ah, I see. All I know is that it did it today but it didn't yesterday 😅

**Comment by pjmtavares** [comment]
Independently of the default value chosen, I think there should be a warning for strings that look like f-strings (contain variable names inside brace pairs) and don't have an f-prefix.

**Reply by pewscorner** [reply]
The warning should be disabled by default, then. Don't want a lot of noise for perfectly valid code.

**Reply by pjmtavares** [reply]
Hum, I guess you use templates/str.format more, and I use f-strings more...

* If the warning would be limited to braces containing names of variables existing in the scope, wouldn't that make it less like to be triggered in your usage?
* Otherwise, if still not good enough, would it make sense to have a setting for this warning like "assume f-strings" / "assume str.format()"?

**Reply by pewscorner** [reply]
I'm not sure what rules would work for every situation out of the box. But I'd be OK with it as long as it is opt-in - or at the very least opt-out.
